test/e2e/adminapi_cluster_getlogs.go

```diff
@@ -0,0 +1,120 @@
+package e2e
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"net/http"
+	"net/url"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+var _ = Describe("[Admin API] Kubernetes get pod logs action", func() {
+	BeforeEach(skipIfNotInDevelopmentEnv)
+
+	const containerName = "e2e-test-container-name"
+	const podName = "e2e-test-pod-name"
+
+	When("in a standard openshift namespace", func() {
+		const namespace = "default"
+
+		It("should be able to get logs from a container of a pod", func() {
+			ctx := context.Background()
+			testGetPodLogsOK(ctx, containerName, podName, namespace)
+		})
+	})
+
+	When("in a customer namespace", func() {
+		const namespace = "e2e-test-namespace"
+
+		It("should be not be able to get logs from customer workload namespaces", func() {
+			ctx := context.Background()
+			testGetPodLogsFromCustomerNamespaceForbidden(ctx, containerName, podName, namespace)
+		})
+	})
+})
+
+// We will create a pod with known logs of its container and will compare the logs gotten through the kubernetes client and through the Admin API.
+func testGetPodLogsOK(ctx context.Context, containerName, podName, namespace string) {
+	By("creating a pod in default namespace with some known logs then comparing those logs with the logs of the container in the pod created")
+	pod := mockPod(containerName, podName, namespace)
+	pod, err := clients.Kubernetes.CoreV1().Pods(namespace).Create(ctx, pod, metav1.CreateOptions{})
+	Expect(err).NotTo(HaveOccurred())
+	defer func() {
+		err = clients.Kubernetes.CoreV1().Pods(namespace).Delete(ctx, pod.Name, metav1.DeleteOptions{})
+		Expect(err).NotTo(HaveOccurred())
+	}()
+	err = wait.Poll(time.Second, time.Second*60, func() (done bool, err error) {
+		pod, err = clients.Kubernetes.CoreV1().Pods(namespace).Get(ctx, pod.Name, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		if pod.Status.Phase == corev1.PodSucceeded {
+			return true, nil
+		}
+
+		return false, nil
+	})
+	Expect(err).NotTo(HaveOccurred())
+
+	params := url.Values{
+		"container": []string{containerName},
+		"namespace": []string{namespace},
+		"podname":   []string{podName},
+	}
+
+	var logs string
+	resp, err := adminRequest(ctx, http.MethodGet, "/admin"+resourceIDFromEnv()+"/kubernetespodlogs", params, nil, &logs)
+	Expect(err).NotTo(HaveOccurred())
+	Expect(resp.StatusCode).To(Equal(http.StatusOK))
+
+	podOptions := corev1.PodLogOptions{
+		Container: pod.Spec.Containers[0].Name,
+	}
+	r := clients.Kubernetes.CoreV1().Pods("default").GetLogs(pod.Name, &podOptions)
+	result, err := r.Do(ctx).Raw()
+	Expect(err).NotTo(HaveOccurred())
+	logs = logs[:len(logs)-1]
```

Comment:
what if we checked that the logs string is not an empty string and error'd out if it was? this would protect us from a panic if the admin action doesn't have the correct response
```suggestion
        if logs == "" || len(logs) < 1 {
          return fmt.Errorf("logs empty string, expected content from admin action")
        }
        logs = logs[:len(logs)-1]
  ```
  you would then need to update the return of the function to take an error.
  
 what do you think?

Reply:
@ross-bryan , The logs are actually what we are defining.(You can check the Pod configuration in func `mockPod` line 109).

> 			
                        Containers: []corev1.Container{{
				Name:    containerName,
				Image:   "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:103505c93bf45c4a29301f282f1ff046e35b63bceaf4df1cca2e631039289da2",
				Command: []string{"/bin/bash", "-c", "echo 'mock-pod-logs'"},
			}},

So the logs are never going to be empty. There is though one possibility where the `logs` can be empty and i.e when there is an err in the following line

>_73_   resp, err := adminRequest(ctx, http.MethodGet, "/admin"+resourceIDFromEnv()+"/kubernetespodlogs", params, nil, &logs)
>_74_   Expect(err).NotTo(HaveOccurred())
And if there is an error, I think it would be checked in the very next line.